File: 443-string-compression/string-compression.cpp
class Solution {
public:
    int compress(vector<char>& chars) {
        int write = 0;  // Position to write in the array
        int read = 0;   // Position to read from the array

        while (read < chars.size()) {
            char current_char = chars[read];
            int count = 0;

            // Count how many times the current character repeats
            while (read < chars.size() && chars[read] == current_char) {
                read++;
                count++;
            }

            // Write the character itself
            chars[write++] = current_char;

            // If the character repeats, write the count
            if (count > 1) {
                for (char digit : to_string(count)) {
                    chars[write++] = digit;
                }
            }
        }

        return write;  // The new length after compression
    }
};

File: 443-string-compression/string-compression.js
/**
 * @param {character[]} chars
 * @return {number}
 */
var compress = function(chars) {

    let a = [];
    let count = 1;

    for(let i = 0; i < chars.length; i++){

        if(chars[i] === chars[i+1]){
            count++;
        } 
        else{
            a.push(chars[i]);

            if(count > 1){
                let str = count.toString();
                for(let c of str){
                    a.push(c);
                }
            }

            count = 1;
        }
    }

    // copy result back into original array
    for(let i = 0; i < a.length; i++){
        chars[i] = a[i];
    }

    return a.length;
};